### PR TITLE
Port the incremental source generator polyfill code to the 6.0 servicing branch.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Backport of: https://github.com/dotnet/runtime/pull/71653 and https://github.com/dotnet/runtime/pull/71651.

Bulk of code added in https://github.com/dotnet/runtime/pull/70911.  However that PR also updated the RegexGenerator (which does not exist in 6.0 and thus does not need to be serviced).

Note: this is a polyfill of an API that roslyn has added.  Once SDK can move to a version of roslyn containing the api, it can reference that API directly and remove the 1000 lines of added code for the polyfill.

### Description:

Primary goal here is to use a more efficient incrementally built up system for determining what classes should be examined to see if we need the source generators to run.  Previously, all types in a solution would be examined (costly in CPU and memory), now only types with attributes on it (that we can show could actually bind to the desired attribute) are checked.  As in practice this is a tiny subset of actual types in a solution (often zero) this vastly decreases the amount of work we needed to be done.

--

Port was very clean, requiring only one deviating line between it and the code that exists in the 7.0 line.

### Customer Impact:

Improved performance of 'in box' generators when in visual studio.  The existing two generators cause lots of CPU and memory churn when processing.  This is because each of them effectively walks all files/types looking for potential matches.  This forces continual reparsing of files, and then lots of semantic binding on potential matches.  For large projects (consider thousands of files being normal), this is an enormous amount of work that can use tons of CPU and memory.  In Roslyn this can be particularly bad, leading to our server getting bogged down and slowing down all language services.

The new approach uses the incremental generator engine to incrementally build up a list of 'actually viable classes' to examine.  The intuition generally being: if you only care about types with "JsonSerializable" on them, then only check types that have that attribute on it syntactically.  This is slightly more involved than just a trivial name check because things like aliases/global-aliases mean that someone could go `global using Foo = System.Text.Json.JsonSerializable;` in one file and then `[Foo] class MyClass` in another.  However, with this, we can narrow down to only the set of types that are actually relevant, not all types.  Furthermore, in projects that reference the sdk but do not use the generator *at all*, there will be zero potential matches, and so no semantic work has to happen at all.

### Regression?

Yes.  This was a perf (but not functionality) regression for anyone with a large solution *and* who moved to .net 6.  Note taht a customer doesn't need to be using the generator to have the perf hit.  Just referencing the SDK is enough as it will cause Roslyn to run the generator, which does all expensive work, just to determine that is has nothing to generate.

### Risk 

I would consider this low.  This is a non-trivial amount of code going in.  However, it has been extensively tested in roslyn, with a large test suite, and has been validated for the 7.0 sdk.  That said, incremental source generation is itself complex (one of the reasons that the perf issue arose in teh first place), so this is definitely more than a trivial fix.  

### Link the PR to the original issue and to the PR to main.

Issue PR:
https://github.com/dotnet/runtime/issues/70754

Main PRs: 
https://github.com/dotnet/runtime/pull/71653
https://github.com/dotnet/runtime/pull/71651
https://github.com/dotnet/runtime/pull/70911

This was multiple PRs as i separated out the initial polyfill, and then did a single PR per generator to update.

### Note any packaging impact. (If the changes touch code in a package that we ship out of band (OOB) then we will need to take care to ensure those packages ship.)

Yes. This impacts the Microsoft.Extensions.Logging.Abstractions NuGet package as well as the System.Text.Json NuGet package. Those two ship source generators that are being edited here so we will need to service those two.

### Note any ref pack impact. (Touches files contained in Microsoft.NETCore.App.Ref, Microsoft.AspNetCore.App.Ref, or Microsoft.WindowsDesktop.App.Ref).

Yes. This impacts the Microsoft.NETCore.App.Ref since the Json Source generator ships in the ref pack as well, and it is being serviced by this change.

### CODE REVIEW

File: src/libraries/Common/src/Roslyn/CSharpSyntaxHelper.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/CSharpSyntaxHelper.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/GetBestTypeByMetadataName.cs [hunk 2]
```diff
@@ -1,7 +1,10 @@
 // Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/GetBestTypeByMetadataName.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/GlobalAliases.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/GlobalAliases.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/Hash.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/Hash.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/ISyntaxHelper.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/ISyntaxHelper.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/SyntaxNodeGrouping.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/SyntaxNodeGrouping.cs:1]: Exact copy of 7.0 line.

File: src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithMetadataName.cs [hunk 7]
```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithMetadataName.cs:1]: Exact copy of 7.0 line.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 8]
```diff
@@ -40,9 +40,10 @@ private sealed class Parser
             private const string JsonPropertyNameAttributeFullName = "System.Text.Json.Serialization.JsonPropertyNameAttribute";
             private const string JsonPropertyOrderAttributeFullName = "System.Text.Json.Serialization.JsonPropertyOrderAttribute";
             private const string JsonSerializerContextFullName = "System.Text.Json.Serialization.JsonSerializerContext";
-            private const string JsonSerializerAttributeFullName = "System.Text.Json.Serialization.JsonSerializableAttribute";
             private const string JsonSourceGenerationOptionsAttributeFullName = "System.Text.Json.Serialization.JsonSourceGenerationOptionsAttribute";
 
+            internal const string JsonSerializableAttributeFullName = "System.Text.Json.Serialization.JsonSerializableAttribute";
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:45]: single deviation.  In 7.0 line this was renamed (from Serializer -> Serializable).  I kept the 7.0 name (more correct), requiring teh update to line 243 below.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs [hunk 9]
```diff
@@ -18,7 +18,7 @@ public partial class LoggerMessageGenerator
     {
```
Comment (CyrusNajmabadi) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:18]: Exact copy of 7.0 line.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Roslyn4.0.cs [hunk 10]
```diff
@@ -7,6 +7,7 @@
 using System.Text;
```
Comment (CyrusNajmabadi) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Roslyn4.0.cs:7]: Exact copy of 7.0 line.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/Microsoft.Extensions.Logging.Generators.Roslyn4.0.csproj [hunk 11]
```diff
@@ -8,6 +8,20 @@
```
Comment (CyrusNajmabadi) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/Microsoft.Extensions.Logging.Generators.Roslyn4.0.csproj:8]: Exact copy of 7.0 line.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs [hunk 12]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.CodeAnalysis;
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:11]: Exact copy of 7.0 line.

File: src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.Roslyn4.0.csproj [hunk 13]
```diff
@@ -13,4 +13,18 @@
     <Compile Include="JsonSourceGenerator.Roslyn4.0.cs" />
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.Roslyn4.0.csproj:13]: Exact copy of 7.0 line.

File: src/libraries/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj [hunk 14]
```diff
@@ -51,6 +50,7 @@
     <Compile Include="$(CommonPath)System\HexConverter.cs" Link="Common\System\HexConverter.cs" />
     <Compile Include="$(CoreLibSharedDir)System\Collections\Generic\ValueListBuilder.cs" Link="Common\System\Collections\Generic\ValueListBuilder.cs" />
     <Compile Include="$(CommonPath)System\Text\ValueStringBuilder.cs" Link="Common\System\Text\ValueStringBuilder.cs" />
+    <Compile Include="$(CommonPath)System\Text\ValueStringBuilder.Pop.cs" Link="Common\System\Text\ValueStringBuilder.Pop.cs" />
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj:53]: I think this was supposed to be ValueListBuilder.pop. I'll fix that on the changes I'm about to push.

**Comment by joperezr** [comment]
The [failure in OSX](https://dev.azure.com/dnceng/public/_build/results?buildId=1881967&view=logs&jobId=0ce95488-260d-5e86-47b9-82bc4e703e0b&j=0ce95488-260d-5e86-47b9-82bc4e703e0b&t=822b0d86-b2cc-5adc-53e4-31a3c632d85c) seems to be related. Looks like an issue with the Logging generator:
```
CSC : error CS8785: Generator 'LoggerMessageGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.' [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]CSC(0,0): error CS8785: (NETCORE_ENGINEERING_TELEMETRY=Build) Generator 'LoggerMessageGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'
CSC : error CS8785: Generator 'LoggerMessageGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.' [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]CSC(0,0): error CS8785: (NETCORE_ENGINEERING_TELEMETRY=Build) Generator 'LoggerMessageGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'
/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(10,32): error CS8795: Partial method 'NoNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(10,32): error CS8795: (NETCORE_ENGINEERING_TELEMETRY=Build) Partial method 'NoNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers.
/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(19,36): error CS8795: Partial method 'OneLevelNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(19,36): error CS8795: (NETCORE_ENGINEERING_TELEMETRY=Build) Partial method 'OneLevelNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers.
/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(31,40): error CS8795: Partial method 'TwoLevelNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MiscTestExtensions.cs(31,40): error CS8795: (NETCORE_ENGINEERING_TELEMETRY=Build) Partial method 'TwoLevelNamespace.CouldNotOpenSocket(ILogger, string)' must have an implementation part because it has accessibility modifiers.
/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/ArgTestExtensions.cs(11,36): error CS8795: Partial method 'ArgTestExtensions.Method1(ILogger)' must have an implementation part because it has accessibility modifiers. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
##[error]src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/ArgTestExtensions.cs(11,36): error CS8795: (NETCORE_ENGINEERING_TELEMETRY=Build) Partial method 'ArgTestExtensions.Method1(ILogger)' must have an implementation part because it has accessibility modifiers.
```

**Comment by joperezr** [comment]
cc: @layomia @eiriktsarpalis @steveharter for the Json Source generator changes.
cc: @eerhardt @maryamariyan for the Logging generator changes.

**Review by eiriktsarpalis — APPROVED** [review]
STJ changes LGTM.

**Review by eerhardt — APPROVED** [review]
Logging generator changes LGTM

**Comment by rbhanda** [comment]
Merge depends on 17.4 Preview 1 testing.

**Comment by tarekgh** [comment]
I labeled it with Meta as it is cover multiple areas.

**Comment by carlossanlop** [comment]
> Merge depends on 17.4 Preview 1 testing.

@rbhanda who should take care of this so I can merge it?

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj [hunk 15]
```diff
@@ -9,7 +9,8 @@
     <Nullable>enable</Nullable>
     <IncludeInternalObsoleteAttribute>true</IncludeInternalObsoleteAttribute>
     <IsPackable>true</IsPackable>
-    <ServicingVersion>5</ServicingVersion>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
```
Comment (carlossanlop) [inline src/libraries/System.Text.Json/src/System.Text.Json.csproj:12]: Thanks for adding the required OOB package authoring changes, here and in M.E.Logging.Abstractions.csproj.

**Comment by carlossanlop** [comment]
> Merge depends on 17.4 Preview 1 testing.

@rbhanda @joperezr @CyrusNajmabadi is this good to merge or do we still need confirmation of the testing? Today is the due date for merging backports that we intend to add to the September servicing release.

**Comment by carlossanlop** [comment]
Tactics approved.
Milestone applied.
CI passed.
OOB package authoring changes included.
Signed off by area owners.
From a chat conversation with @joperezr, we didn't hear about any testing issues, so we're good.
Ready to merge! :shipit:

**Pull request merged**